### ISSUE TITLE
App container startup issue if startup depends on network connectivity via sidecar proxy

### ISSUE THREAD

**Issue by shashankram** [openservicemesh/osm#2316]
**Please describe the Improvement and/or Feature Request**
If an app container startup depends on network to be ready, the app container startup will fail until the sidecar proxy container is running and has received all configuration to allow network traffic.

If the pod's `RestartPolicy` is set to `Never`, the app container will die and not start again. If the `RestartPolicy` is set to `Always`, the app container will keep crashing till network connectivity is ready. This is a problem because in either case users are going to wonder why their app container crashed. With the existing code, an app container that relies on networking to be ready on startup is guaranteed to crash.

Example:
The `mysql-client` container below relies on connectivity to a mysql server to be ready on startup, else the container will crash
```
kubectl run -it --rm --image=mysql:5.6 --restart=Never mysql-client -- mysql -h mysql -ppassword
```

**Scope (please mark with X where applicable)**
- New Functionality          [X]
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- CI System                  [ ]
- Project Release            [ ]

**Possible use cases**
The use case is to be able to reliably start an app container whose startup depends on networking to be ready.

Following are some of the possibilities to solve this problem:
1. Provide CNI support and startup the sidecar proxy via the CNI
2. Wrap application using [envoy-preflight](https://github.com/monzo/envoy-preflight) to ensure an app container only starts up after envoy proxy is ready
3. Implement in OSM a workaround to only start app container after the sidecar proxy starts up using a post-startup hook for the sidecar container. Reference: https://medium.com/@marko.luksa/delaying-application-start-until-sidecar-is-ready-2ec2d21a7b74

**Comment by draychev** [comment]
Thank you for researching this issue, @shashankram! This is very interesting!
I referenced it in the v0.8 planning issue https://github.com/openservicemesh/osm/issues/2168 so we pick it up in v0.8 (or ever earlier if possible)

**Comment by steeling** [comment]
I'll add, similar to waiting for startup, is waiting for shut down. This is particularly useful for jobs, that only reach completion when *all* containers stop. However the sidecar never stops (but *is* injected to pods backed by jobs).

This is closer to a bug and may bring the priority of this issue up higher

**Comment by nojnhuh** [comment]
Overall, I'm inclined to say the issue regarding startup is better solved by workloads than by OSM. 

This same problem is theoretically possible with just Kubernetes without OSM if, for example, the endpoints controller is backed up for whatever reason (since Kubernetes informers also operate on an eventual consistency model) where a Service might exist but Endpoints have not yet been set to route to the right Pods. A workload making a request to that service at startup in that scenario could also crash.

And this problem isn't unique to startup. Envoy's configuration can drift out of sync with what's correct given the state of the world at any time, so requests might fail for similar reasons well after startup. It's entirely possible those requests could also cause the workload to crash. There's some overlap here with #3044. The idea there involved building strict consistency on top of Envoy's eventual consistency model, which seemed like it was going to cause lots of friction.

I think it will be much simpler overall if workloads handle this themselves where they can control the exact mechanism used to wait (polling/sleeping/etc.) and timeouts or other parameters. Or they can choose not to handle this at all if it's not necessary. While workloads shouldn't have to be aware that they're operating in the context of Envoy or Kubernetes specifically, I think it's reasonable to expect they're aware they're operating in an environment primarily driven by eventual consistency where not every operation is expected to succeed at all times.

**Comment by steeling** [comment]
I agree with Jon. It might be useful to provide documentation (although it exists externally), such that when people search for this issue with OSM they find our recommended solutions

**Comment by nojnhuh** [comment]
Regarding the shutdown issue, I've done some exploratory work in #4841 to make OSM stop sidecars when Jobs are done. Based on that, it seems like the overall simplest, safest, and most flexible workaround until there is first-class support for sidecars in Kubernetes is for Jobs in the mesh to stop the sidecars themselves by calling Envoy's `/quitquitquit` admin endpoint.

For containers running shell scripts, this might involve adding a line like
```sh
trap 'curl localhost:15000/quitquitquit -X POST' EXIT
```

Or if curl isn't available but busybox is,
```sh
trap 'nc 127.0.0.1 15000 -e echo -ne "POST /quitquitquit HTTP/1.1\nHost: 127.0.0.1:15000\n\n"' EXIT
```

I'm not sure where this info might fit best on the docs site, but maybe this issue is a good enough spot for now?
